x-pack/plugins/maps/public/shared/layers/sources/client_file_source/geojson_file_source.js

```diff
@@ -0,0 +1,120 @@
+import uuid from 'uuid/v4';
+import _ from 'lodash';
+
+export class GeojsonFileSource extends AbstractVectorSource {
+
+  static type = GEOJSON_FILE;
+  static title = 'Import GeoJSON Vector File';
+  static description = 'Upload a GeoJSON file & Index in Elasticsearch';
+  static icon = 'emsApp';
+  static indexReadyFile = true;
+
+  static createDescriptor(featureCollection, name) {
+    return {
+      type: GeojsonFileSource.type,
+      featureCollection,
+      name
+    };
+  }
+
+  static viewIndexedData = (addAndViewSource, inspectorAdapters, onIndexSuccess) => {
+    return ({ fields, id, success }) => {
+      if (!success) {
+        console.error('Unable to view indexed data');
+        return;
+      }
+      const geoFieldArr = fields.filter(
+        field => ['geo_point', 'geo_shape'].includes(field.type)
+      );
+      const geoField = _.get(geoFieldArr, '[0].name');
+      const indexPatternId = id;
+      if (!indexPatternId || !geoField) {
+        addAndViewSource(null);
+      } else {
+        const source = new ESSearchSource({
+          id: uuid(),
+          indexPatternId,
+          geoField,
+        }, inspectorAdapters);
+        addAndViewSource(source);
+        onIndexSuccess();
+      }
+    };
+  };
+
+  static previewGeojsonFile = (onPreviewSource, inspectorAdapters) => {
+    return (geojsonFile, name) => {
+      if (!geojsonFile) {
+        onPreviewSource(null);
+        return;
+      }
+      const sourceDescriptor = GeojsonFileSource.createDescriptor(geojsonFile, name);
+      const source = new GeojsonFileSource(sourceDescriptor, inspectorAdapters);
+      onPreviewSource(source);
+    };
+  };
+
+  static renderEditor({ onPreviewSource, inspectorAdapters, addAndViewSource,
+    boolIndexData, onRemove, onIndexReadyStatusChange, setFailures, onIndexSuccess }) {
+    return (
+      <ClientFileCreateSourceEditor
+        previewGeojsonFile={
+          GeojsonFileSource.previewGeojsonFile(
+            onPreviewSource,
+            inspectorAdapters
+          )
+        }
+        viewIndexedData={
+          GeojsonFileSource.viewIndexedData(
+            addAndViewSource,
+            inspectorAdapters,
+            onIndexSuccess
+          )
+        }
+        boolIndexData={boolIndexData}
+        onRemove={onRemove}
+        onIndexReadyStatusChange={onIndexReadyStatusChange}
+        onIndexAddSuccess={indexAddResp => {
+          if (indexAddResp.failures && indexAddResp.failures.length) {
+            setFailures && setFailures(indexAddResp.failures);
+          }
+        }}
+      />
+    );
+  }
+
+  async getGeoJsonWithMeta() {
+    return {
+      data: this._descriptor.featureCollection,
+      meta: {}
+    };
+  }
+
+  async getDisplayName() {
+    try {
+      return this._descriptor.name;
+    } catch (error) {
+      return this._descriptor.id;
+    }
+  }
+
+  canFormatFeatureProperties() {
+    return true;
+  }
+
+  shouldBeIndexed() {
```

Comment:
this is a little bizar in how this works. So all implementations access the `indexReadyFile` static property. so to implement this correctly GeoJsonSource needs to both override the property AND this method.

Just retain the `shouldBeIndexed` getter, and have the top level in `AbstractSource` return `false`, and this on return `true`. That static property is never accessed directly from the class anyway.

Reply:
It was written as you're describing, but we needed this information prior to instantiating an instance when we're selecting sources. It's used [here](https://github.com/elastic/kibana/blob/99632b301168d2baa9ed9605a752e81efe798d30/x-pack/plugins/maps/public/components/layer_addpanel/view.js#L66).